src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java

```diff
@@ -196,71 +199,95 @@ private static Input parseArguments(String[] args) {
     public static void main(String[] args) throws Exception {
         Input input = parseArguments(args);
 
-        FormatToFileNamesWrapper formatToFileNamesWrapper = new FormatToFileNamesWrapper();
+        ReportPatternManager reportPatternManager = new ReportPatternManager();
 
-        formatToFileNamesWrapper.addReportFilePatternsFromInputFile(input.input, input.format);
-        formatToFileNamesWrapper.addFilePatternsForFormat(input.files, input.format);
+        if (input.inputFile != null) {
+            reportPatternManager.addReportFilePatternsFromInputFile(input.inputFile);
+        }
+        reportPatternManager.addFilePatternsForFormat(input.files, input.format);
 
-        performUpload(formatToFileNamesWrapper, input);
+        OkHttpClient client = OkHttpClientUtils.createClient(input.validateSsl);
+        try {
+            performUpload(client, reportPatternManager, input);
+        } finally {
+            // we must shut down OkHttp as otherwise it will leave threads running and
+            // prevent JVM shutdown
+            client.dispatcher().executorService().shutdownNow();
+            client.connectionPool().evictAll();
+        }
     }
 
-    private static void performUpload(FormatToFileNamesWrapper formatToFileNamesWrapper, Input input) throws IOException, AgentOptionParseException {
-        String sessionId = openSession(input);
-        for (String format : formatToFileNamesWrapper.getAllAvailableFormats()) {
-            sendRequestForFormat(input, format, formatToFileNamesWrapper.getFilesForFormat(format), sessionId);
+    private static void performUpload(OkHttpClient client, ReportPatternManager reportPatternManager, Input input) throws IOException, AgentOptionParseException {
+        String sessionId = openSession(client, input);
+        for (String format : reportPatternManager.getAllUsedFormats()) {
+            Set<String> patternsForFormat = reportPatternManager.getPatternsForFormat(format);
+            if (patternsForFormat.isEmpty()) {
```

Comment:
can we move this validation step into the `PatternManager`? Otherwise there's a possibility that we upload 9/10 formats and then the 10th fails and the session is never closed. In that case I'd prefer to fail early during input file parsing instead.